fairscale/optim/layerwise_gradient_scaler.py

```diff
@@ -0,0 +1,278 @@
+import logging
+from typing import List, Tuple
+
+import torch
+import torch.nn as nn
+
+
+class LayerInfo:
+    """
+    A class to record the layer attributes.
+    """
+
+    def __init__(self, name: str, layer: nn.Module, scale: float = 1.0, scale_layer: bool = False) -> None:
+        """
+        layer_name: name of the layer e.g. fc1, conv1, relu1
+        layer_type: type of the layer e.g. Linear, Conv2d, ReLU
+        scaling_factor: user configurable scaling factor for the layer, defaults to 1.0
+        found_inf_or_nan: a boolean indicating if any parameter of layers' gradient contains inf/nan
+        growth_tracker: tracks number of step since last time scale was increased
+        scale_layer: a boolean indicating if the layer should be scaled or not
+        """
+        self.layer_name = name
+        self.layer_type = layer
+        self.scaling_factor = scale
+        self.found_inf_or_nan = False
+        self.growth_tracker = 0
+        self.scale_layer = scale_layer
+
+
+class GradientHelper:
+    """
+    A helper class to create instances of backward hooks. The hooks are registered in the
+    scale method of LayerwiseGradientScaler.
+    """
+
+    def __init__(self, name: str, inputs_multiplier: float, outputs_multiplier: float):
+        self.layer_name = name
+        self.inputs_multiplier = inputs_multiplier
+        self.outputs_multiplier = outputs_multiplier
+
+    def scale_gradients(self, m: nn.Module, inputs: Tuple, outputs: Tuple) -> Tuple[torch.Tensor]:
+        """
+        Backward hook that is attached to the layers to scale the gradients.
+        """
+        scaled_up_grads = list()
+        for idx in range(len(inputs)):
+            if inputs[idx] is not None:
+                if self.inputs_multiplier != 1.0 or self.outputs_multiplier != 1.0:
+                    logging.debug(
+                        "layer = %s \t scale = %s \t scale_down = %s"
+                        % (self.layer_name, self.inputs_multiplier, self.outputs_multiplier)
+                    )
+                scaled_up_grads.append(inputs[idx].mul(self.inputs_multiplier * self.outputs_multiplier))
+            else:
+                logging.debug("next layer is None")
+                scaled_up_grads.append(inputs[idx])
+        return tuple(scaled_up_grads)  # type: ignore
+
+
+class LayerwiseGradientScaler:
+    """
+    LayerwiseGradientScaler enables using distinct scaling factors for each layer
+    of the network.
+
+    Example:
+
+    # Create a convolutional network
+        class ConvNet(nn.Module):
+            def __init__(self):
+                ...
+
+            def forward(self, x):
+                ...
+
+        # Create an instance of the model
+        model = ConvNet()
+        optimizer = torch.optim.SGD(model.parameters())
+
+        # specify the layers to scale and their scaling factor
+        layer_scale_dict = {"conv1": 2**10, "conv2": 2**8, "fc1": 2**10, "fc2": 2**9}
```

Comment:
(Not something to implement in this PR but just a thought)
What happens if the number of layers are in the order of 10s or 100s? Should there be difference scaling factors applied in terms of depth i.e the first 25% are scale factor 1, next are scale factor 2? I suspect @MidoAssran 's model may not be that large hence you may not need it immediately. Just something we should think about from a API perspective for use cases we may need to support. WDYT?

Reply:
In deep models the scaling factors will have to be set carefully by the user. Initializing all layers with the same scaling factor may not produce the best results. After training and observing the gradient overflow we can determine which layers require gradient scaling. The common practice is to initialize scaling factor to a large value and let the Gradient Scaler unscale the scaling factor so that it comes into a reasonable range.